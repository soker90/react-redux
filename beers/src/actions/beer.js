import beerAction from '../constants/actions/beer'

export default {
  fetchBeer: () => ({type: beerAction.FETCH_BEER}),
  fetchBeerSuccess: beers => ({type: beerAction.FETCH_BEER_SUCCESS, beers}),
  addBeer: (name, graduation, price) => ({type: beerAction.ADD_BEER, name, graduation, price}),
  addBeerSuccess: beer => ({type: beerAction.ADD_BEER_SUCCESS, beer}),
  removeBeer: beerId => ({type: beerAction.REMOVE_BEER, beerId}),
  removeBeerSuccess: beerId => ({type: beerAction.REMOVE_BEER_SUCCESS, beerId}),
  beerError: error => ({type: beerAction.BEER_ERROR, error})
}
