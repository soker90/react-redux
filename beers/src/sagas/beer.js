import { call, put, takeEvery } from 'redux-saga/effects'
import addBeerService from '../services/addBeer'
import fetchBeerService from '../services/fetchBeer'
import removeBeerService from '../services/removeBeer'
import beerActions from '../actions/beer'
import beerActionTypes from '../constants/actions/beer'

export function * fetchBeer () {
  try {
    const response = yield call(fetchBeerService)
    yield put(beerActions.fetchBeerSuccess(response))
  } catch (error) {
    yield put(beerActions.beerError(error))
  }
}

export function * watchFetchBeer () {
  yield takeEvery(beerActionTypes.FETCH_BEER, fetchBeer)
}

export function * addBeer (beer) {
  try {
    const response = yield call(addBeerService, beer)
    yield put(beerActions.addBeerSuccess(response))
  } catch (error) {
    yield put(beerActions.beerError(error))
  }
}

export function * watchAddBeer () {
  yield takeEvery(beerActionTypes.ADD_BEER, addBeer)
}

export function * removeBeer (beerId) {
  try {
    yield call(removeBeerService, beerId)
    yield put(beerActions.removeBeerSuccess(beerId))
  } catch (error) {
    yield put(beerActions.beerError(error))
  }
}

export function * watchRemoveBeer () {
  yield takeEvery(beerActionTypes.REMOVE_BEER, removeBeer)
}
