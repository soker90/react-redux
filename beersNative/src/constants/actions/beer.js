export default {
  FETCH_BEER: 'FETCH_BEER',
  FETCH_BEER_SUCCESS: 'FETCH_BEER_SUCCESS',
  ADD_BEER: 'ADD_BEER',
  ADD_BEER_SUCCESS: 'ADD_BEER_SUCCESS',
  REMOVE_BEER: 'REMOVE_BEER',
  REMOVE_BEER_SUCCESS: 'REMOVE_BEER_SUCCESS',
  BEER_ERROR: 'BEER_ERROR'
}
