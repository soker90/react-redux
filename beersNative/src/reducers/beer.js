import beerAction from '../constants/actions/beer'

export function initialState () {
  return {
    beerList: [],
    beer: {},
    loading: false,
    error: null
  }
}

function fetchBeer(state) {
  return {
    ...state,
    beerList: [],
    loading: true,
    error: null
  }
}

function fetchBeerSuccess(state, {beers}) {
  return {
    ...state,
    beerList: beers.beers,
    loading: false,
    error: null
  }
}

function addBeer(state) {
  return {
    ...state,
    loading: true,
    error: null
  }
}

function addBeerSuccess(state, {beer}) {
  return {
    ...state,
    beerList: [beer, ...state.beerList],
    loading: false,
    error: null
  }
}

function removeBeer(state) {
  return {
    ...state,
    loading: true,
    error: null
  }
}

function removeBeerSuccess(state, {beerId}) {
  return {
    ...state,
    beerList: state.beerList.filter((beer) => beer.id !== beerId.beerId),
    loading: false,
    error: null
  }
}

function beerError(state, {error}) {
  return {
    ...state,
    loading: false,
    error: error
  }
}

export default function (state = initialState(), action) {
  switch (action.type) {
    case beerAction.FETCH_BEER:
      return fetchBeer(state, action)
    case beerAction.FETCH_BEER_SUCCESS:
      return fetchBeerSuccess(state, action)
    case beerAction.ADD_BEER:
      return addBeer(state, action)
    case beerAction.ADD_BEER_SUCCESS:
      return addBeerSuccess(state, action)
    case beerAction.REMOVE_BEER:
      return removeBeer(state, action)
    case beerAction.REMOVE_BEER_SUCCESS:
      return removeBeerSuccess(state, action)
    case beerAction.BEER_ERROR:
      return beerError(state, action)
    default:
      return state
  }
}
