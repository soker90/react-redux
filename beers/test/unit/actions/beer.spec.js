// beer.spec.js
import beer from '../../../src/actions/beer'

describe('Beer action', () => {
  test('fetchBeer', () => {
    expect(beer.fetchBeer()).toMatchSnapshot()
  })
  test('fetchBeerSuccess', () => {
    expect(beer.fetchBeerSuccess()).toMatchSnapshot()
  })
  test('addBeer', () => {
    expect(beer.addBeer()).toMatchSnapshot()
  })
  test('addBeerSuccess', () => {
    expect(beer.addBeerSuccess()).toMatchSnapshot()
  })
  test('removeBeer', () => {
    expect(beer.removeBeer()).toMatchSnapshot()
  })
  test('removeBeerSuccess', () => {
    expect(beer.removeBeerSuccess()).toMatchSnapshot()
  })
  test('beerError', () => {
    expect(beer.beerError()).toMatchSnapshot()
  })
})
