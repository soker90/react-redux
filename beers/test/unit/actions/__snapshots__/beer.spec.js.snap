exports[`Beer action addBeer 1`] = `
Object {
  "graduation": undefined,
  "name": undefined,
  "type": "ADD_BEER",
}
`;

exports[`Beer action addBeerSuccess 1`] = `
Object {
  "beer": undefined,
  "type": "ADD_BEER_SUCCESS",
}
`;

exports[`Beer action beerError 1`] = `
Object {
  "error": undefined,
  "type": "BEER_ERROR",
}
`;

exports[`Beer action fetchBeer 1`] = `
Object {
  "type": "FETCH_BEER",
}
`;

exports[`Beer action fetchBeerSuccess 1`] = `
Object {
  "beers": undefined,
  "type": "FETCH_BEER_SUCCESS",
}
`;

exports[`Beer action removeBeer 1`] = `
Object {
  "beerId": undefined,
  "type": "REMOVE_BEER",
}
`;

exports[`Beer action removeBeerSuccess 1`] = `
Object {
  "beerId": undefined,
  "type": "REMOVE_BEER_SUCCESS",
}
`;
